src/LinuxCaptureService/LinuxCaptureService.cpp

```diff
@@ -113,17 +113,58 @@ static void StopInternalProducersAndCaptureStartStopListenersInParallel(
   }
 }
 
+namespace {
+
+// This class hijacks FunctionEntry and FunctionExit events before they reach the
+// ProducerEventProcessor, and sends them to LinuxTracing instead, so that they can be processed
+// like u(ret)probes. All the other events are forwarded to the ProducerEventProcessor normally.
+class ProducerEventProcessorHijackingFunctionEntryExitForLinuxTracing
+    : public ProducerEventProcessor {
+ public:
+  ProducerEventProcessorHijackingFunctionEntryExitForLinuxTracing(
+      ProducerEventProcessor* original_producer_event_processor, TracingHandler* tracing_handler)
+      : producer_event_processor_{original_producer_event_processor},
+        tracing_handler_{tracing_handler} {}
+  ~ProducerEventProcessorHijackingFunctionEntryExitForLinuxTracing() override = default;
+
+  void ProcessEvent(uint64_t producer_id,
+                    orbit_grpc_protos::ProducerCaptureEvent&& event) override {
+    switch (event.event_case()) {
+      case ProducerCaptureEvent::kFunctionEntry:
+        tracing_handler_->ProcessFunctionEntry(event.function_entry());
+        break;
+      case ProducerCaptureEvent::kFunctionExit:
+        tracing_handler_->ProcessFunctionExit(event.function_exit());
+        break;
+      case ProducerCaptureEvent::EVENT_NOT_SET:
+        UNREACHABLE();
+      default:
+        producer_event_processor_->ProcessEvent(producer_id, std::move(event));
+    }
+  }
+
+ private:
+  ProducerEventProcessor* producer_event_processor_;
+  TracingHandler* tracing_handler_;
+};
+
+}  // namespace
+
 grpc::Status LinuxCaptureService::Capture(
     grpc::ServerContext* /*context*/,
     grpc::ServerReaderWriter<CaptureResponse, CaptureRequest>* reader_writer) {
   orbit_base::SetCurrentThreadName("CSImpl::Capture");
 
-  grpc::Status result = InitializeCapture(reader_writer);
-  if (!result.ok()) {
-    return result;
+  {
```

Comment:
Is there a reason to put this in an extra scope (fine for me though).

Reply:
I didn't want to leak `result` below. For example there's another `result` used in another scope below.
But I just realized I can declare it in the if condition directly (I know you don't like that too much but I think it's better than the extra scope).

Reply:
I'm fine with this :)